components/script/script_thread.rs

```diff
@@ -1649,15 +1649,26 @@ impl ScriptThread {
         // TODO: The specification says to filter out non-renderable documents,
         // as well as those for which a rendering update would be unnecessary,
         // but this isn't happening here.
-        let pipeline_and_docs: Vec<(PipelineId, DomRoot<Document>)> = self
+        let pipelines_to_update: Vec<PipelineId> = self
             .documents
             .borrow()
             .iter()
-            .map(|(id, document)| (id, DomRoot::from_ref(&*document)))
+            .filter(|(_, document)| document.window().is_top_level())
+            .flat_map(|(id, document)| {
+                std::iter::once(id.clone()).chain(
```

Comment:
@gterzian I tried to used `iter::once` here with `chain' to avoid vec allocation.